### PR TITLE
add soundex function

### PR DESCRIPTION
**Pull request by bestgopher**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Summary about this PR

## Changelog

- New Feature
 

## Related Issues

Fixes #issue

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by bestgopher** [comment]
#3045

File: common/functions/src/scalars/strings/soundex.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
+        match columns[0].data_type() {
+            DataType::String => {
+                return match columns[0].column().cast_with_type(&DataType::String) {
```
Comment (sundy-li) [inline common/functions/src/scalars/strings/soundex.rs:136]: Already String, no need to cast

Comment (bestgopher) [reply common/functions/src/scalars/strings/soundex.rs:136]: Yeah. I will fix it at once.

File: common/functions/src/scalars/strings/soundex.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                return match columns[0].column() {
+                    DataColumn::Constant(DataValue::String(Some(ref data)), _) => {
+                        let r = Self::soundex(data);
+                        Ok(DataColumn::Constant(DataValue::String(Some(r)), input_rows))
+                    }
+                    _ => Ok(DataColumn::Constant(DataValue::Null, input_rows)),
+                };
+            }
+            DataType::Null => Ok(DataColumn::Constant(DataValue::Null, input_rows)),
+            _ => {
```
Comment (sundy-li) [inline common/functions/src/scalars/strings/soundex.rs:145]: I did not know why this branch returns an empty String column.

Maybe we should cast the column into String column in previous.

```

// Null/Primitive array can cast into String column.
let column = columns[0].cast_with_type(DataType::String);
let column = column.string()?;

match column {
   Constant => {},
   Array(x) => {}
}

```

File: common/functions/src/scalars/strings/soundex.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        &*self._display_name
+    }
+
+    fn num_arguments(&self) -> usize {
+        1
+    }
+
+    fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+        match args[0] {
+            DataType::Null => Ok(DataType::Null),
```
Comment (sundy-li) [inline common/functions/src/scalars/strings/soundex.rs:124]:
```
  if args[0].is_numeric() || args[0] == DataType::String || args[0] == DataType::Null {
            Ok(DataType::String)
        } else {
            Err(ErrorCode::IllegalDataType(format!(
                "Expected string/numeric, but got {}",
                args[0]
            )))
        }
```

**Comment by BohuTANG** [comment]
@mergify update

**Comment by BohuTANG** [comment]
Thank you @bestgopher

**Pull request merged**